### ISSUE TITLE
MainTest is too slow for Travis

### ISSUE THREAD

**Issue by monperrus** [INRIA/spoon#2662]
MainTest sometimes takes more than 10 minutes without any output and the build fails because of this, see https://travis-ci.org/INRIA/spoon/jobs/440920311

```
[INFO] Running spoon.test.main.MainTest
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

**Comment by pvojtechovsky** [comment]
How long takes MainTest on CI usually? ... when it pass...
Is it near to 10 min or much less?

**Comment by monperrus** [comment]
Close to 10 minutes.

**Comment by monperrus** [comment]
In essence, MainTest is not a unit test. It's a parametrized test that is meant to hold for all correct Java code in the world.

As such, it is not part of standard CI.

Here is a proposal:

* we rename `MainTest` into `ContractVerifier`, and put it into `src/main/java`, this would enable to run it on any application: `java -jar spoon.jar -verify -i your-favorite-code/src`
* in level-1 CI (Travis) we only run `ContractVerifier` on all classes that are test resources (`**.testclasses.**`). 
* in level-2 CI (Jenkins),  we run `ContractVerifier` on all projects we already have (Guava, etc.)

WDYT?

**Comment by surli** [comment]
OK for extracting the test in a ContractVerifier and to run it on Jenkins. I don't really understand why it's interesting to run it on Travis on `**.testclasses.**`.
Looking back the MainTest it was only triggered on `src/main/java`.
So I'd push it on jenkins and I'd launch it on `src/main/java`.

Side note: your ContractVerifier looks really close to the concept of ModelValidator I exposed in #2701. Could be interesting to refactor it in that way in a near future.

**Comment by nharrand** [comment]
> I don't really understand why it's interesting to run it on Travis on `**.testclasses.**`.

I think the point of this, is (in the end) to generate a up to date minified set of classes by tracing level-2 CI, and let level-1 CI run only on this set. Like that level-1 CI still provides non regression verification, but remain fast. But level-2 still provide a more exhaustive validation.

**Comment by monperrus** [comment]
@nharrand is correct:

- we don't want to completely give up ContractVerifier in Travis
- yet we want it to be fast
- so we would run it on a small set of Java files

`**.testclasses.**` is only a first attempt 1) that is small enough 2) that should be OK in fullclasspath (no dependency outside). 

Future work will surely improve this subset.

**Comment by surli** [comment]
>  2) that should be OK in fullclasspath (no dependency outside).

Don't agree, I'm pretty sure we have imports of some classes from Spoon sources. Remember we have test testclasses with annotations, with processors, etc.

I understand the need to run it on a smallsubset, I'm just not sure `**.testclasses.**` is the best choice for that.

**Comment by monperrus** [comment]
fixed by #2716
